include/pybind11/pythoncapi_compat.h

```diff
@@ -0,0 +1,346 @@
+// Header file providing new functions of the Python C API to old Python
+// versions.
+//
+// File distributed under the MIT license.
+//
+// Homepage:
+// https://github.com/pythoncapi/pythoncapi_compat
+//
+// Latest version:
+// https://raw.githubusercontent.com/pythoncapi/pythoncapi_compat/master/pythoncapi_compat.h
+//
+// SPDX-License-Identifier: MIT
+
+#ifndef PYTHONCAPI_COMPAT
+#define PYTHONCAPI_COMPAT
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <Python.h>
```

Comment:
Another thought/concern: here we have manipulations that need to happen *before* Python.h is included:

https://github.com/pybind/pybind11/blob/1259db6fd99bc60228ef0e4af2e379f8e86abc1d/include/pybind11/detail/common.h#L105

First idea: replace the existing `#include <Python.h>` in `common.h` with including this header.

Reply:
There are actually several reasons to control Python.h's include order. By the way, note on the header - you shouldn't need extern C when including Python.h, it explicitly supports inclusion from C++.